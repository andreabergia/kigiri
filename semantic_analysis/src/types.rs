use parser::LiteralValue;
use std::fmt::Display;
use std::fmt::Formatter;

#[derive(Debug, PartialEq, Clone, Copy)]
pub enum Type {
    Int,
    Double,
    Boolean,
}

impl Display for Type {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Type::Int => write!(f, "int"),
            Type::Double => write!(f, "double"),
            Type::Boolean => write!(f, "boolean"),
        }
    }
}

impl Type {
    pub fn to_string_short(&self) -> String {
        match self {
            Type::Int => "i".to_string(),
            Type::Double => "d".to_string(),
            Type::Boolean => "b".to_string(),
        }
    }

    pub fn of_literal(literal: &LiteralValue) -> Self {
        match literal {
            LiteralValue::Integer(_) => Type::Int,
            LiteralValue::Double(_) => Type::Double,
            LiteralValue::Boolean(_) => Type::Boolean,
        }
    }
}
