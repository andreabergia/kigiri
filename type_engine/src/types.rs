use std::fmt::Display;
use std::fmt::Formatter;

#[derive(Debug, PartialEq, Clone)]
pub enum Type {
    Int,
    Double,
    Boolean,
}

impl Display for Type {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Type::Int => write!(f, "int"),
            Type::Double => write!(f, "double"),
            Type::Boolean => write!(f, "boolean"),
        }
    }
}

impl Type {
    pub fn to_string_short(&self) -> String {
        match self {
            Type::Int => "i".to_string(),
            Type::Double => "d".to_string(),
            Type::Boolean => "b".to_string(),
        }
    }
}
