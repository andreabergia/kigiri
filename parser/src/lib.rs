mod ast;
mod grammar;
mod memory;
mod parser;
