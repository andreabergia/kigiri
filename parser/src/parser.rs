use crate::ast::{
    BinaryOperator, Block, Expression, FunctionArgument, FunctionDeclaration, LetInitializer,
    Module, Statement, UnaryOperator,
};
use crate::grammar::{Grammar, Rule};
use crate::parsed_ast::{FunctionSignaturesByName, ParsedAstAllocator, PhaseParsed};
use crate::symbols::intern_string;
use bumpalo::collections::Vec as BumpVec;
use pest::Parser;
use pest::iterators::Pair;
use pest::pratt_parser::{Assoc, Op, PrattParser};
use std::str::FromStr;
use std::sync::LazyLock;

static PRATT_PARSER: LazyLock<PrattParser<Rule>> = LazyLock::new(|| {
    PrattParser::new()
        .op(Op::infix(Rule::or, Assoc::Left))
        .op(Op::infix(Rule::and, Assoc::Left))
        .op(Op::infix(Rule::bitwise_and, Assoc::Left)
            | Op::infix(Rule::bitwise_or, Assoc::Left)
            | Op::infix(Rule::bitwise_xor, Assoc::Left))
        .op(Op::infix(Rule::eq, Assoc::Left) | Op::infix(Rule::neq, Assoc::Left))
        .op(Op::infix(Rule::lt, Assoc::Left)
            | Op::infix(Rule::lte, Assoc::Left)
            | Op::infix(Rule::gt, Assoc::Left)
            | Op::infix(Rule::gte, Assoc::Left))
        .op(Op::infix(Rule::bitwise_shl, Assoc::Left) | Op::infix(Rule::bitwise_shr, Assoc::Left))
        .op(Op::infix(Rule::add, Assoc::Left) | Op::infix(Rule::sub, Assoc::Left))
        .op(Op::infix(Rule::mul, Assoc::Left)
            | Op::infix(Rule::div, Assoc::Left)
            | Op::infix(Rule::rem, Assoc::Left))
        .op(Op::infix(Rule::exp, Assoc::Right))
        .op(Op::prefix(Rule::neg) | Op::prefix(Rule::not) | Op::prefix(Rule::bitwise_not))
});

fn parse_expression<'a>(
    ast_allocator: &'a ParsedAstAllocator,
    rule: Pair<'_, Rule>,
) -> &'a Expression<'a, PhaseParsed<'a>> {
    PRATT_PARSER
        .map_primary(|primary| match primary.as_rule() {
            Rule::number => {
                let pair = primary.into_inner().next().unwrap();
                let text = pair.as_str();
                match pair.as_rule() {
                    Rule::intNumber => ast_allocator.literal_int(i64::from_str(text).unwrap()),
                    Rule::hexNumber => ast_allocator.literal_int(
                        i64::from_str_radix(text.to_ascii_lowercase().trim_start_matches("0x"), 16)
                            .unwrap(),
                    ),
                    Rule::doubleNumber => {
                        ast_allocator.literal_double(f64::from_str(text).unwrap())
                    }
                    _ => unreachable!(""),
                }
            }
            Rule::identifier => ast_allocator.identifier(primary.as_str()),
            Rule::expression => parse_expression(ast_allocator, primary),
            Rule::boolean => ast_allocator.literal_boolean(primary.as_str().parse().unwrap()),
            Rule::functionCall => parse_function_call(ast_allocator, primary),
            _ => unreachable!(""),
        })
        .map_prefix(|prefix, operand| match prefix.as_rule() {
            Rule::neg => ast_allocator.unary(UnaryOperator::Neg, operand),
            Rule::not => ast_allocator.unary(UnaryOperator::Not, operand),
            Rule::bitwise_not => ast_allocator.unary(UnaryOperator::BitwiseNot, operand),
            _ => unreachable!(),
        })
        .map_infix(|left, op, right| match op.as_rule() {
            Rule::add => ast_allocator.binary(BinaryOperator::Add, left, right),
            Rule::mul => ast_allocator.binary(BinaryOperator::Mul, left, right),
            Rule::sub => ast_allocator.binary(BinaryOperator::Sub, left, right),
            Rule::div => ast_allocator.binary(BinaryOperator::Div, left, right),
            Rule::rem => ast_allocator.binary(BinaryOperator::Rem, left, right),
            Rule::exp => ast_allocator.binary(BinaryOperator::Exp, left, right),
            Rule::eq => ast_allocator.binary(BinaryOperator::Eq, left, right),
            Rule::neq => ast_allocator.binary(BinaryOperator::Neq, left, right),
            Rule::lt => ast_allocator.binary(BinaryOperator::Lt, left, right),
            Rule::lte => ast_allocator.binary(BinaryOperator::Lte, left, right),
            Rule::gt => ast_allocator.binary(BinaryOperator::Gt, left, right),
            Rule::gte => ast_allocator.binary(BinaryOperator::Gte, left, right),
            Rule::and => ast_allocator.binary(BinaryOperator::And, left, right),
            Rule::or => ast_allocator.binary(BinaryOperator::Or, left, right),
            Rule::bitwise_and => ast_allocator.binary(BinaryOperator::BitwiseAnd, left, right),
            Rule::bitwise_or => ast_allocator.binary(BinaryOperator::BitwiseOr, left, right),
            Rule::bitwise_xor => ast_allocator.binary(BinaryOperator::BitwiseXor, left, right),
            Rule::bitwise_shl => ast_allocator.binary(BinaryOperator::BitwiseShl, left, right),
            Rule::bitwise_shr => ast_allocator.binary(BinaryOperator::BitwiseShr, left, right),
            _ => unreachable!(),
        })
        .parse(rule.into_inner())
}

fn parse_function_call<'a>(
    ast_allocator: &'a ParsedAstAllocator,
    rule: Pair<'_, Rule>,
) -> &'a Expression<'a, PhaseParsed<'a>> {
    let mut inner = rule.into_inner();
    let name = inner.next().unwrap().as_str();

    let mut args = ast_allocator.function_call_arguments();
    for arg in inner.next().unwrap().into_inner() {
        args.push(parse_expression(ast_allocator, arg));
    }

    ast_allocator.function_call(name, args)
}

fn parse_let_statement<'a>(
    ast_allocator: &'a ParsedAstAllocator,
    rule: Pair<'_, Rule>,
) -> &'a Statement<'a, PhaseParsed<'a>> {
    let mut iter = rule.into_inner();
    let mut initializers = ast_allocator.statement_let_initializers();
    loop {
        let Some(initializer_rule) = iter.next() else {
            break;
        };
        let mut initializer_rule = initializer_rule.into_inner();

        let id = initializer_rule.next().unwrap();
        let id = intern_string(id.as_str());

        let value = initializer_rule.next().unwrap();
        let value = parse_expression(ast_allocator, value);

        initializers.push(LetInitializer {
            variable: id,
            value,
        })
    }
    ast_allocator.statement_let(initializers)
}

fn parse_statement<'a>(
    ast_allocator: &'a ParsedAstAllocator,
    rule: Pair<'_, Rule>,
) -> &'a Statement<'a, PhaseParsed<'a>> {
    let pair = rule.into_inner().next().unwrap();
    match pair.as_rule() {
        Rule::letStatement => parse_let_statement(ast_allocator, pair),
        Rule::assignmentStatement => {
            let mut inner = pair.into_inner();
            let identifier = inner
                .next()
                .expect("identifier on lhs of assignment")
                .as_str();
            let expression = parse_expression(
                ast_allocator,
                inner.next().expect("expression on rhs of assignment"),
            );
            ast_allocator.statement_assignment(identifier, expression)
        }
        Rule::returnStatement => {
            let expression = pair
                .into_inner()
                .next()
                .map(|p| parse_expression(ast_allocator, p));
            ast_allocator.statement_return(expression)
        }
        Rule::expression => {
            let expression = parse_expression(ast_allocator, pair);
            ast_allocator.statement_expression(expression)
        }
        _ => unreachable!(),
    }
}

fn parse_block<'a>(
    ast_allocator: &'a ParsedAstAllocator,
    rule: Pair<'_, Rule>,
) -> &'a Block<'a, PhaseParsed<'a>> {
    // We want a parent block to have a smaller id than any nested block,
    // so we generate the block_id first and then recurse.
    let block_id = ast_allocator.next_block_id();

    let mut statements = ast_allocator.statements();
    for pair in rule.into_inner() {
        match pair.as_rule() {
            Rule::statement => statements.push(parse_statement(ast_allocator, pair)),
            Rule::block => {
                statements.push(ast_allocator.nested_block(parse_block(ast_allocator, pair)))
            }
            _ => unreachable!(),
        }
    }
    ast_allocator.block_from_statements(block_id, statements)
}

fn parse_function_declaration_arguments<'a>(
    ast_allocator: &'a ParsedAstAllocator,
    rule: Pair<'_, Rule>,
) -> BumpVec<'a, FunctionArgument> {
    let mut arguments = ast_allocator.function_arguments();

    for arg in rule.into_inner() {
        let mut arg = arg.into_inner();
        let name = arg.next().expect("argument name").as_str();
        let arg_type = arg.next().expect("argument type").as_str();
        arguments.push(FunctionArgument {
            name: intern_string(name),
            argument_type: intern_string(arg_type),
        })
    }

    arguments
}

fn parse_function_declaration<'a>(
    ast_allocator: &'a ParsedAstAllocator,
    rule: Pair<'_, Rule>,
) -> &'a FunctionDeclaration<'a, PhaseParsed<'a>> {
    let mut rule_iter = rule.into_inner();
    let name = rule_iter.next().expect("function name").as_str();

    let arguments_rule = rule_iter.next().expect("function arguments");
    let arguments = parse_function_declaration_arguments(ast_allocator, arguments_rule);

    let next = rule_iter.next().expect("function return type or body");
    let (return_type, body_rule) = if let Rule::functionReturnType = next.as_rule() {
        let return_type = next.into_inner().next().expect("return type").as_str();
        (
            Some(intern_string(return_type)),
            rule_iter.next().expect("function body"),
        )
    } else {
        (None, next)
    };

    let body = parse_block(ast_allocator, body_rule);

    assert!(rule_iter.next().is_none());

    ast_allocator.function_declaration(name, return_type, arguments, body)
}

fn parse_module<'a>(
    ast_allocator: &'a ParsedAstAllocator,
    module_name: &str,
    rule: Pair<'_, Rule>,
) -> &'a Module<'a, PhaseParsed<'a>> {
    let mut functions = ast_allocator.functions();
    let mut function_signatures = FunctionSignaturesByName::default();

    let rule_inner = rule.into_inner();
    for inner in rule_inner.into_iter() {
        match inner.as_rule() {
            Rule::EOI => continue,
            Rule::functionDeclaration => {
                let fun = parse_function_declaration(ast_allocator, inner);
                function_signatures.insert(fun.signature.name, fun.signature);
                functions.push(fun);
            }
            _ => unreachable!(),
        }
    }

    ast_allocator.module(module_name, functions, function_signatures)
}

pub fn parse_as_expression<'a>(
    ast_allocator: &'a ParsedAstAllocator,
    text: &str,
) -> &'a Expression<'a, PhaseParsed<'a>> {
    let pair = Grammar::parse(Rule::expression, text)
        .unwrap()
        .next()
        .unwrap();
    parse_expression(ast_allocator, pair)
}

pub fn parse_as_block<'a>(
    ast_allocator: &'a ParsedAstAllocator,
    text: &str,
) -> &'a Block<'a, PhaseParsed<'a>> {
    let pair = Grammar::parse(Rule::block, text).unwrap().next().unwrap();
    parse_block(ast_allocator, pair)
}

pub fn parse<'a>(
    ast_allocator: &'a ParsedAstAllocator,
    module_name: &str,
    text: &str,
) -> &'a Module<'a, PhaseParsed<'a>> {
    let pair = Grammar::parse(Rule::module, text).unwrap().next().unwrap();
    parse_module(ast_allocator, module_name, pair)
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Generates a test case to verify the AST produced by a given source expression.
    /// The AST is passed as its string representation.
    /// The macro allows us to one-line a test case, reducing noise such as the
    /// `#[test]` annotation and the various newline required by a function.
    macro_rules! test_expression {
        ($name: ident, $source: expr, $ast: expr) => {
            #[test]
            fn $name() {
                let ast_allocator = ParsedAstAllocator::default();
                let expression = parse_as_expression(&ast_allocator, $source);
                assert_eq!(expression.to_string(), $ast);
            }
        };
    }

    macro_rules! test_block {
        ($name: ident, $source: expr, $ast: expr) => {
            #[test]
            fn $name() {
                let ast_allocator = ParsedAstAllocator::default();
                let block = parse_as_block(&ast_allocator, $source);
                assert_eq!(block.to_string(), $ast);
            }
        };
    }

    test_expression!(literal_int_1, "1", "1i");
    test_expression!(literal_int_2, "3217832", "3217832i");
    test_expression!(literal_int_3, "0xA1", "161i");

    test_expression!(literal_float_1, "0.", "0d");
    test_expression!(literal_float_2, "3.14", "3.14d");
    test_expression!(literal_float_3, "1.1e+2", "110d");
    test_expression!(literal_float_4, "10e-1", "1d");
    test_expression!(literal_float_5, ".1e1", "1d");
    test_expression!(literal_float_6, "1e4", "10000d");

    test_expression!(identifier_1, "i", "i");
    test_expression!(identifier_2, "a_b", "a_b");
    test_expression!(identifier_3, "éñò", "éñò");
    test_expression!(identifier_4, "_a", "_a");
    test_expression!(identifier_5, "_", "_");

    test_expression!(boolean_1, "true", "true");
    test_expression!(boolean_2, "false", "false");

    test_expression!(precedence_01, "1 + 2 * 3", "(+ 1i (* 2i 3i))");
    test_expression!(precedence_02, "1 - 2 / 3", "(- 1i (/ 2i 3i))");
    test_expression!(precedence_03, "1 + 2 % 3", "(+ 1i (% 2i 3i))");
    test_expression!(precedence_04, "1 * 2 ** 3", "(* 1i (** 2i 3i))");
    test_expression!(precedence_05, "1 == 2 + 3", "(== 1i (+ 2i 3i))");
    test_expression!(precedence_06, "1 != 2 + 3", "(!= 1i (+ 2i 3i))");
    test_expression!(precedence_07, "1 < 2 + 3", "(< 1i (+ 2i 3i))");
    test_expression!(precedence_08, "1 <= 2 + 3", "(<= 1i (+ 2i 3i))");
    test_expression!(precedence_09, "1 > 2 + 3", "(> 1i (+ 2i 3i))");
    test_expression!(precedence_10, "1 >= 2 + 3", "(>= 1i (+ 2i 3i))");
    test_expression!(precedence_11, "1 && 2 == 3", "(&& 1i (== 2i 3i))");
    test_expression!(precedence_12, "1 || 2 && 3", "(|| 1i (&& 2i 3i))");
    test_expression!(precedence_13, "1 & 2 + 3", "(& 1i (+ 2i 3i))");
    test_expression!(precedence_14, "1 + 2 | 3", "(| (+ 1i 2i) 3i)");
    test_expression!(precedence_15, "1 ^ 2 == 3", "(^ 1i (== 2i 3i))");
    test_expression!(precedence_16, "1 | 2 && 3", "(&& (| 1i 2i) 3i)");
    test_expression!(precedence_17, "1 + 2 << 3", "(<< (+ 1i 2i) 3i)");
    test_expression!(precedence_18, "1 == 2 >> 3", "(== 1i (>> 2i 3i))");
    test_expression!(precedence_19, "1 == 2 > 3", "(== 1i (> 2i 3i))");
    test_expression!(precedence_20, "1 != 2 <= 3", "(!= 1i (<= 2i 3i))");
    test_expression!(precedence_21, "1 + - 2", "(+ 1i (- 2i))");
    test_expression!(precedence_22, "1 + ~ 2", "(+ 1i (~ 2i))");
    test_expression!(precedence_23, "1 && ! 2", "(&& 1i (! 2i))");

    test_expression!(parenthesis, "(1 + 2) * 3", "(* (+ 1i 2i) 3i)");

    test_expression!(function_call_no_args, "foo()", "foo()");
    test_expression!(function_call_one_arg, "foo(1)", "foo(1i)");
    test_expression!(
        function_call_multiple_args,
        "foo(1, 2, 3)",
        "foo(1i, 2i, 3i)"
    );

    test_block!(
        statement_expression,
        r"{
   1;
}",
        r"{ #0
1i;
}"
    );
    test_block!(
        statement_return_with_value,
        r"{
   return 1;
}",
        r"{ #0
return 1i;
}"
    );
    test_block!(
        statement_return_empty,
        r"{
   return;
}",
        r"{ #0
return;
}"
    );
    test_block!(
        statement_let_initializer,
        r"{
   let a = 1;
}",
        r"{ #0
let a = 1i;
}"
    );
    test_block!(
        statement_let_multiple_initializers,
        r"{
   let a = 1, b = false;
}",
        r"{ #0
let a = 1i, b = false;
}"
    );
    test_block!(
        statement_assignment,
        r"{
   a = 1;
}",
        r"{ #0
a = 1i;
}"
    );
    test_block!(
        statement_nested_block,
        r"{
       {
          a = 1;
       }
    }",
        r"{ #0
{ #1
a = 1i;
}
}"
    );

    #[test]
    fn module_fn_with_return_type() {
        let ast_allocator = ParsedAstAllocator::default();
        let module = parse(
            &ast_allocator,
            "test",
            r"
fn add(x: int, y: int) -> int {
    return x + y;
}
",
        );
        assert_eq!(
            module.to_string(),
            r"module test

fn add(
    x: int,
    y: int,
) -> int
{ #0
return (+ x y);
}
"
        );
    }

    #[test]
    fn module_fn_no_return_type() {
        let ast_allocator = ParsedAstAllocator::default();
        let module = parse(
            &ast_allocator,
            "test",
            r"
fn foo() {}
",
        );
        assert_eq!(
            module.to_string(),
            r"module test

fn foo(
) -> void
{ #0
}
"
        );
    }
}
