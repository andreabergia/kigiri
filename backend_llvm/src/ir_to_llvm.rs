use codegen::{InstructionId, InstructionPayload, LiteralValue};
use inkwell::builder::{Builder, BuilderError};
use inkwell::context::Context;
use inkwell::module::Module;
use inkwell::types::FunctionType;
use inkwell::values::IntValue;
use inkwell::IntPredicate;
use parser::{resolve_string_id, BinaryOperator, UnaryOperator};
use semantic_analysis::Type;
use thiserror::Error;

#[derive(Debug, PartialEq, Error)]
#[error("Code generation error: {message}")]
struct CodeGenError {
    message: String,
}

impl From<BuilderError> for CodeGenError {
    fn from(value: BuilderError) -> Self {
        Self {
            message: value.to_string(),
        }
    }
}

struct FunctionContext<'c> {
    // Vectors are indexed by instruction id. There's a bit of space wasted,
    // but it makes everything quite simple and fast.
    int_values: Vec<Option<IntValue<'c>>>,
    bool_values: Vec<Option<IntValue<'c>>>,
}

impl FunctionContext<'_> {
    fn new(num_instructions: usize) -> Self {
        let mut int_values = Vec::with_capacity(num_instructions);
        int_values.resize(num_instructions, None);

        let mut bool_values = Vec::with_capacity(num_instructions);
        bool_values.resize(num_instructions, None);

        Self {
            int_values,
            bool_values,
        }
    }
}

struct LlvmGenerator<'c, 'm, 'm2>
where
    'c: 'm,
{
    context: &'c Context,
    llvm_module: Module<'c>,
    builder: Builder<'c>,
    ir_module: &'m codegen::Module<'m2>,
}

impl<'c, 'm, 'm2> LlvmGenerator<'c, 'm, 'm2> {
    fn new(context: &'c Context, ir_module: &'m codegen::Module<'m2>) -> Self {
        let llvm_module =
            context.create_module(resolve_string_id(ir_module.name).expect("module name"));
        let builder = context.create_builder();
        Self {
            context,
            llvm_module,
            builder,
            ir_module,
        }
    }

    fn generate(&mut self) -> Result<(), CodeGenError> {
        for function in self.ir_module.functions.iter() {
            self.generate_fun(function)?;
        }
        Ok(())
    }

    fn generate_fun(&mut self, function: &codegen::Function) -> Result<(), CodeGenError> {
        let mut fun_ctx = FunctionContext::new(function.body.instructions.borrow().len());

        let fn_type = self.make_fun_type(function);
        let fun = self.llvm_module.add_function(
            resolve_string_id(function.signature.name).expect("function name"),
            fn_type,
            None,
        );

        let bb = self.context.append_basic_block(fun, "entry");
        self.builder.position_at_end(bb);

        for instruction in function.body.instructions.borrow().iter() {
            match &instruction.payload {
                InstructionPayload::Constant { constant, .. } => {
                    self.handle_constant(&mut fun_ctx, instruction.id, constant);
                }

                InstructionPayload::Unary {
                    operand_type,
                    operator,
                    operand,
                } => {
                    self.handle_unary(
                        &mut fun_ctx,
                        instruction.id,
                        operand_type,
                        operator,
                        *operand,
                    )?;
                }
                InstructionPayload::Binary {
                    operand_type,
                    operator,
                    left,
                    right,
                } => {
                    self.handle_binary(
                        &mut fun_ctx,
                        instruction.id,
                        operand_type,
                        operator,
                        *left,
                        *right,
                    )?;
                }
                InstructionPayload::Ret => {
                    self.builder.build_return(None)?;
                }
                &InstructionPayload::RetExpr { .. } => todo!(),
                &InstructionPayload::Load { .. } => todo!(),
            }
        }

        if !fun.verify(true) {
            panic!("Invalid function");
        }

        fun.print_to_stderr();
        Ok(())
    }

    fn make_fun_type(&mut self, function: &codegen::Function) -> FunctionType<'c> {
        let arguments = function
            .signature
            .arguments
            .iter()
            .map(|arg| match arg.argument_type {
                Type::Int => self.context.i64_type().into(),
                Type::Boolean => self.context.bool_type().into(),
                Type::Double => self.context.f64_type().into(),
            })
            .collect::<Vec<_>>();

        match &function.signature.return_type {
            None => self.context.void_type().fn_type(&arguments, false),
            Some(t) => match t {
                Type::Int => self.context.i64_type().fn_type(&arguments, false),
                Type::Boolean => self.context.bool_type().fn_type(&arguments, false),
                Type::Double => self.context.f64_type().fn_type(&arguments, false),
            },
        }
    }

    fn handle_binary(
        &mut self,
        fun_ctx: &mut FunctionContext<'c>,
        id: InstructionId,
        operand_type: &Type,
        operator: &BinaryOperator,
        left: InstructionId,
        right: InstructionId,
    ) -> Result<(), CodeGenError> {
        match operand_type {
            Type::Int => {
                self.handle_binary_int(fun_ctx, id, operator, left, right)?;
            }

            Type::Boolean => {
                self.handle_binary_boolean(fun_ctx, id, operator, left, right)?;
            }
            Type::Double => todo!(),
        }
        Ok(())
    }

    fn handle_binary_boolean(
        &mut self,
        fun_ctx: &mut FunctionContext<'c>,
        id: InstructionId,
        operator: &BinaryOperator,
        left: InstructionId,
        right: InstructionId,
    ) -> Result<(), CodeGenError> {
        let left = fun_ctx
            .bool_values
            .get(left.as_usize())
            .expect("vector should be initialized with the correct length")
            .expect("binary operand should be a bool");
        let right = fun_ctx
            .bool_values
            .get(right.as_usize())
            .expect("vector should be initialized with the correct length")
            .expect("binary operand should be a bool");

        fun_ctx.bool_values[id.as_usize()] = Some(match operator {
            BinaryOperator::Add
            | BinaryOperator::Sub
            | BinaryOperator::Mul
            | BinaryOperator::Div
            | BinaryOperator::Rem
            | BinaryOperator::Exp
            | BinaryOperator::Lt
            | BinaryOperator::Lte
            | BinaryOperator::Gt
            | BinaryOperator::Gte
            | BinaryOperator::BitwiseAnd
            | BinaryOperator::BitwiseOr
            | BinaryOperator::BitwiseXor
            | BinaryOperator::BitwiseShl
            | BinaryOperator::BitwiseShr => unreachable!(),

            BinaryOperator::Eq => self.builder.build_int_compare(
                IntPredicate::EQ,
                left,
                right,
                &Self::name("eq", id),
            )?,
            BinaryOperator::Neq => self.builder.build_int_compare(
                IntPredicate::NE,
                left,
                right,
                &Self::name("neq", id),
            )?,
            BinaryOperator::And => self
                .builder
                .build_and(left, right, &Self::name("and", id))?,
            BinaryOperator::Or => self.builder.build_or(left, right, &Self::name("or", id))?,
        });
        Ok(())
    }

    fn handle_binary_int(
        &mut self,
        fun_ctx: &mut FunctionContext<'c>,
        id: InstructionId,
        operator: &BinaryOperator,
        left: InstructionId,
        right: InstructionId,
    ) -> Result<(), CodeGenError> {
        let left = fun_ctx
            .int_values
            .get(left.as_usize())
            .expect("vector should be initialized with the correct length")
            .expect("binary operand should be an int");
        let right = fun_ctx
            .int_values
            .get(right.as_usize())
            .expect("vector should be initialized with the correct length")
            .expect("binary operand should be an int");

        fun_ctx.int_values[id.as_usize()] = Some(match operator {
            BinaryOperator::Add => {
                self.builder
                    .build_int_add(left, right, &Self::name("add", id))?
            }
            BinaryOperator::Sub => {
                self.builder
                    .build_int_sub(left, right, &Self::name("sub", id))?
            }
            BinaryOperator::Mul => {
                self.builder
                    .build_int_mul(left, right, &Self::name("mul", id))?
            }
            BinaryOperator::Div => {
                self.builder
                    .build_int_signed_div(left, right, &Self::name("div", id))?
            }
            BinaryOperator::Rem => {
                todo!()
            }
            BinaryOperator::Exp => {
                todo!()
            }
            BinaryOperator::Eq => self.builder.build_int_compare(
                IntPredicate::EQ,
                left,
                right,
                &Self::name("eq", id),
            )?,
            BinaryOperator::Neq => self.builder.build_int_compare(
                IntPredicate::NE,
                left,
                right,
                &Self::name("neq", id),
            )?,
            BinaryOperator::Lt => self.builder.build_int_compare(
                IntPredicate::SLT,
                left,
                right,
                &Self::name("lt", id),
            )?,
            BinaryOperator::Lte => self.builder.build_int_compare(
                IntPredicate::SLE,
                left,
                right,
                &Self::name("lte", id),
            )?,
            BinaryOperator::Gt => self.builder.build_int_compare(
                IntPredicate::SGT,
                left,
                right,
                &Self::name("gt", id),
            )?,
            BinaryOperator::Gte => self.builder.build_int_compare(
                IntPredicate::SGE,
                left,
                right,
                &Self::name("gte", id),
            )?,
            BinaryOperator::BitwiseAnd => {
                self.builder
                    .build_and(left, right, &Self::name("bitwise_and", id))?
            }
            BinaryOperator::BitwiseOr => {
                self.builder
                    .build_or(left, right, &Self::name("bitwise_or", id))?
            }
            BinaryOperator::BitwiseXor => {
                self.builder
                    .build_xor(left, right, &Self::name("bitwise_xor", id))?
            }
            BinaryOperator::BitwiseShl => {
                self.builder
                    .build_left_shift(left, right, &Self::name("bitwise_shl", id))?
            }
            BinaryOperator::BitwiseShr => self.builder.build_right_shift(
                left,
                right,
                false,
                &Self::name("bitwise_shr", id),
            )?,

            BinaryOperator::And | BinaryOperator::Or => {
                // TODO
                unreachable!()
            }
        });
        Ok(())
    }

    fn handle_unary(
        &mut self,
        fun_ctx: &mut FunctionContext<'c>,
        id: InstructionId,
        operand_type: &Type,
        operator: &UnaryOperator,
        operand: InstructionId,
    ) -> Result<(), CodeGenError> {
        match operand_type {
            Type::Int => {
                self.handle_unary_int(fun_ctx, id, operator, operand)?;
            }
            Type::Boolean => {
                self.handle_unary_boolean(fun_ctx, id, operator, operand)?;
            }
            Type::Double => todo!(),
        }
        Ok(())
    }

    fn name(prefix: &str, id: InstructionId) -> String {
        format!("{}_{}", prefix, id)
    }

    fn handle_unary_boolean(
        &mut self,
        fun_ctx: &mut FunctionContext<'c>,
        id: InstructionId,
        operator: &UnaryOperator,
        operand: InstructionId,
    ) -> Result<(), CodeGenError> {
        let operand = fun_ctx
            .bool_values
            .get(operand.as_usize())
            .expect("vector should be initialized with the correct length")
            .expect("unary operand should be a bool");
        fun_ctx.bool_values[id.as_usize()] = Some(match operator {
            UnaryOperator::Neg | UnaryOperator::BitwiseNot => {
                unreachable!()
            }
            UnaryOperator::Not => self.builder.build_not(operand, &Self::name("not", id))?,
        });
        Ok(())
    }

    fn handle_unary_int(
        &mut self,
        fun_ctx: &mut FunctionContext<'c>,
        id: InstructionId,
        operator: &UnaryOperator,
        operand: InstructionId,
    ) -> Result<(), CodeGenError> {
        let operand = fun_ctx
            .int_values
            .get(operand.as_usize())
            .expect("vector should be initialized with the correct length")
            .expect("unary operand should be an int");
        fun_ctx.int_values[id.as_usize()] = Some(match operator {
            UnaryOperator::Neg => self
                .builder
                .build_int_neg(operand, &Self::name("neg", id))?,
            UnaryOperator::BitwiseNot => {
                // There's no LLVM instruction for bitwise not, but we can xor with
                // all 1s to get the same result (i.e. complement to 1)
                self.builder.build_xor(
                    self.context.i64_type().const_all_ones(),
                    operand,
                    &Self::name("bitwise_not", id),
                )?
            }
            UnaryOperator::Not => {
                // TODO: error
                unreachable!("unexpected not operator with int type")
            }
        });
        Ok(())
    }

    fn handle_constant(
        &mut self,
        fun_ctx: &mut FunctionContext<'c>,
        id: InstructionId,
        constant: &LiteralValue,
    ) {
        match constant {
            LiteralValue::Integer(value) => {
                let option = fun_ctx
                    .int_values
                    .get_mut(id.as_usize())
                    .expect("vector should be initialized with the correct length");
                *option = Some(self.context.i64_type().const_int(*value as u64, false));
            }
            LiteralValue::Boolean(value) => {
                let option = fun_ctx
                    .bool_values
                    .get_mut(id.as_usize())
                    .expect("vector should be initialized with the correct length");
                *option = Some(
                    self.context
                        .bool_type()
                        .const_int(if *value { 1 } else { 0 }, false),
                );
            }
            LiteralValue::Double(..) => todo!(),
        };
    }
}

#[allow(unused)]
fn ir_to_llvm(context: &Context, module: &codegen::Module) -> Result<(), CodeGenError> {
    let mut builder = LlvmGenerator::new(context, module);
    builder.generate()
}

#[cfg(test)]
mod tests {
    use super::*;
    use codegen::build_ir_module;
    use codegen::IrAllocator;
    use inkwell::context::Context;
    use semantic_analysis::{SemanticAnalyzer, TypedModule};
    use std::io::{stderr, Write};

    // TODO: this needs to not be so duplicated across projects
    fn make_analyzed_ast<'s>(
        semantic_analyzer: &'s SemanticAnalyzer,
        source: &str,
    ) -> &'s TypedModule<'s> {
        let ast_allocator = parser::AstAllocator::default();
        let module = parser::parse(&ast_allocator, "test", source);

        let result = semantic_analyzer.analyze_module(module);
        result.expect("should have passed semantic analysis")
    }

    fn handle_module<'i>(ir_allocator: &'i IrAllocator, source: &str) -> &'i codegen::Module<'i> {
        let semantic_analyzer = SemanticAnalyzer::default();
        let module = make_analyzed_ast(&semantic_analyzer, source);
        let module = build_ir_module(ir_allocator, module);

        eprintln!("Module IR:\n{}", module);
        stderr().flush().unwrap();

        module
    }

    #[test]
    fn test_ir_to_llvm() {
        let ir_allocator = IrAllocator::new();
        let basic_block = handle_module(
            &ir_allocator,
            r"fn simple() {
  1 + 2;
}
",
            // fn add_one(x: int) -> int {
            //     return 1;
            // }
            // ",
        );

        let context = Context::create();
        ir_to_llvm(&context, basic_block).unwrap();
    }
}
