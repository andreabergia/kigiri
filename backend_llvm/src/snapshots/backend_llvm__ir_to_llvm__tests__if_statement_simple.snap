---
source: backend_llvm/src/ir_to_llvm.rs
expression: llvm_ir
---
; ModuleID = 'test'
source_filename = "test"

define i64 @test(i1 %x) {
bb0:
  br i1 %x, label %bb2, label %bb1

bb2:                                              ; preds = %bb0
  ret i64 1

bb1:                                              ; preds = %bb0
  ret i64 0
}
